zebra/zebra_nb_state.c

```diff
@@ -155,235 +170,414 @@ lib_interface_zebra_state_mcast_group_get_elem(const char *xpath,
 	return yang_data_new_ipv4(xpath, &vxlan_info->mcast_grp);
 }
 
-const void *lib_vrf_ribs_rib_get_next(const void *parent_list_entry,
-				      const void *list_entry)
+const void *lib_vrf_zebra_ribs_rib_get_next(const void *parent_list_entry,
+					    const void *list_entry)
 {
-	/* TODO: implement me. */
-	return NULL;
+	struct vrf *vrf = (struct vrf *)parent_list_entry;
+	struct zebra_router_table *zrt =
+		(struct zebra_router_table *)list_entry;
+
+	struct zebra_vrf *zvrf;
+	afi_t afi;
+	safi_t safi;
+
+	zvrf = zebra_vrf_lookup_by_id(vrf->vrf_id);
+
+	if (list_entry == NULL) {
+		afi = AFI_IP;
+		safi = SAFI_UNICAST;
+
+		/* Start from AFI_IP and SAFI_UNICAST */
+		zrt = zebra_router_find_zrt(zvrf, zvrf->table_id, afi, safi);
+		if (zrt == NULL)
+			return NULL;
+	} else {
+		zrt = RB_NEXT(zebra_router_table_head, zrt);
+		/* vrf_id/ns_id do not match, only walk for the given VRF */
+		while (zrt && zrt->ns_id != zvrf->zns->ns_id)
+			zrt = RB_NEXT(zebra_router_table_head, zrt);
+	}
+
+	if (IS_ZEBRA_DEBUG_EVENT && zrt)
+		zlog_debug("%s: vrf %s afi %s safi %s", __func__, vrf->name,
+			   afi2str(zrt->afi), safi2str(zrt->safi));
+
+	return zrt;
 }
 
-int lib_vrf_ribs_rib_get_keys(const void *list_entry,
-			      struct yang_list_keys *keys)
+int lib_vrf_zebra_ribs_rib_get_keys(const void *list_entry,
+				    struct yang_list_keys *keys)
 {
-	/* TODO: implement me. */
+	const struct zebra_router_table *zrt = list_entry;
+
+	assert(zrt);
+
+	keys->num = 2;
+	zebra_extract_afi_safi_key(keys->key[0], sizeof(keys->key[0]), zrt->afi,
+				   zrt->safi, "frr-zebra");
+
+	snprintf(keys->key[1], sizeof(keys->key[1]), "%" PRIu32, zrt->tableid);
+
+	if (IS_ZEBRA_DEBUG_EVENT)
+		zlog_debug("%s: key[0] %s key[1] %s", __func__, keys->key[0],
```

Comment:
Please remove all debugs from the state callbacks. `debug northbound callbacks state` should take care of doing that in a generic way for us.

Unfortunately the NB doesn't display the list keys in the debug messages, but that's something that can be easily implemented later.

Reply:
I have removed most of the debugs but it is evidently useful to have keys printed in many cases where keys are derived via munching.